sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisReader.java

```diff
@@ -0,0 +1,146 @@
+ */
+package org.apache.beam.sdk.io.kinesis;
+
+
+import org.apache.beam.sdk.io.UnboundedSource;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.collect.Lists.newArrayList;
+
+import org.joda.time.Instant;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import java.io.IOException;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+
+/***
+ * Reads data from multiple kinesis shards in a single thread.
+ */
+class KinesisReader extends UnboundedSource.UnboundedReader<KinesisRecord> {
+    private static final Logger LOG = LoggerFactory.getLogger(KinesisReader.class);
+
+    private final SimplifiedKinesisClient kinesis;
+    private final UnboundedSource<KinesisRecord, ?> source;
+    private final CheckpointGenerator initialCheckpointGenerator;
+    private RoundRobin<ShardRecordsIterator> shardIterators;
+    private CustomOptional<KinesisRecord> currentRecord = CustomOptional.absent();
+
+    public KinesisReader(SimplifiedKinesisClient kinesis,
+                         CheckpointGenerator initialCheckpointGenerator,
+                         UnboundedSource<KinesisRecord, ?> source) {
+        checkNotNull(kinesis);
+        checkNotNull(initialCheckpointGenerator);
+
+        this.kinesis = kinesis;
+        this.source = source;
+        this.initialCheckpointGenerator = initialCheckpointGenerator;
+    }
+
+    /***
+     * Generates initial checkpoint and instantiates iterators for shards.
+     */
+    @Override
+    public boolean start() throws IOException {
+        LOG.info("Starting reader using {}", initialCheckpointGenerator);
+
+        try {
+            KinesisReaderCheckpoint initialCheckpoint =
+                    initialCheckpointGenerator.generate(kinesis);
+            List<ShardRecordsIterator> iterators = newArrayList();
+            for (ShardCheckpoint checkpoint : initialCheckpoint) {
+                iterators.add(checkpoint.getShardRecordsIterator(kinesis));
+            }
+            shardIterators = new RoundRobin<>(iterators);
+        } catch (TransientKinesisException e) {
+            throw new IOException(e);
+        }
+
+        return advance();
+    }
+
+    /***
+     * Moves to the next record in one of the shards.
+     * If current shard iterator can be move forward (i.e. there's a record present) then we do it.
+     * If not, we iterate over shards in a round-robin manner.
+     */
+    @Override
+    public boolean advance() throws IOException {
+        try {
+            for (int i = 0; i < shardIterators.size(); ++i) {
+                currentRecord = shardIterators.getCurrent().next();
+                if (currentRecord.isPresent()) {
+                    return true;
+                } else {
+                    shardIterators.moveForward();
+                }
+            }
+        } catch (TransientKinesisException e) {
+            LOG.warn("Transient exception occurred: {}", e);
```

Comment:
no `{}` necessary on logs like this: slf4j will automatically recognize exception class and do the right thing